include/deal.II/fe/fe_values.h

```diff
@@ -5429,6 +5507,43 @@ FEValuesBase<dim, spacedim>::get_inverse_jacobians() const
 
 
 
+template <int dim, int spacedim>
+inline boost::integer_range<unsigned int>
+FEValuesBase<dim, spacedim>::local_dof_indices() const
+{
+  return boost::irange(0U, dofs_per_cell);
+}
+
+
+
+template <int dim, int spacedim>
+inline boost::integer_range<unsigned int>
+FEValuesBase<dim, spacedim>::local_dof_indices_starting_at(
+  const unsigned int start_dof_index) const
+{
+  Assert(
+    start_dof_index <= dofs_per_cell,
```

Comment:
I'll recheck this definition and that of `local_dof_indices_ending_at`. This was a late night effort, so I've likely made a mistake here (although the test passes, which is a super small plus).

Reply:
> 

Permitting an empty range is useful. Consider copying one have of a symmetric matrix to another. This mimics how we have done it in a tutorial:
```
    for (const unsigned int i : scratch.fe_values_ref.dof_indices())
      for (const unsigned int j :
           scratch.fe_values_ref.dof_indices_starting_at(i + 1))
        data.cell_matrix(i, j) = data.cell_matrix(j, i);
```
We want to start the indices at `i+1` so that we don't wastefully copy the diagonal onto itself. However, in the very last row we just have the diagonal entry to iterate over so this would then go out of bounds if we didn't allow an empty range.

Does this seem reasonable to you?